orb-ui/src/engine/operator/pulse.rs

```diff
@@ -63,38 +67,67 @@ impl Animation for Pulse {
         dt: f64,
         idle: bool,
     ) -> AnimationState {
-        if let Some(phase) = self.phase.as_mut() {
-            if *phase >= self.solid_period && self.wave_period != 0.0 {
-                *phase += dt * (PI * 2.0 / self.wave_period);
-            } else {
-                *phase += dt;
+        if self.wave_period == 0.0 {
+            return AnimationState::Finished;
+        } else if let Some(Transition::ForceStop) = self.transition {
+            self.wave_period = 0.0;
+            return AnimationState::Finished;
+        } else if let Some(Transition::StartDelay(duration)) = self.transition {
+            self.transition_time += dt;
+            if self.transition_time >= duration {
+                self.transition = None;
             }
-            *phase %= PI * 2.0 + self.solid_period;
-            if !idle {
-                let color = if *phase >= self.solid_period {
-                    let intensity = if self.inverted {
-                        // starts at intensity 0
-                        (1.0 - (*phase - self.solid_period).cos()) / 2.0
-                    } else {
-                        // starts at intensity 1
-                        ((*phase - self.solid_period).cos() + 1.0) / 2.0
-                    };
-                    self.color * intensity
+            return AnimationState::Running;
+        }
+
+        if self.phase >= self.solid_period && self.wave_period != 0.0 {
+            self.phase += dt * (PI * 2.0 / self.wave_period);
+        } else {
+            self.phase += dt;
+        }
+        self.phase %= PI * 2.0 + self.solid_period;
+        if !idle {
+            let color = if self.phase >= self.solid_period {
+                let intensity = if self.inverted {
+                    // starts at intensity 0
+                    (1.0 - (self.phase - self.solid_period).cos()) / 2.0
                 } else {
-                    // solid
-                    self.color
+                    // starts at intensity 1
+                    ((self.phase - self.solid_period).cos() + 1.0) / 2.0
                 };
-                for led in frame {
-                    *led = color;
+                if let Some(Transition::PlayOnce) = self.transition {
+                    if intensity < 0.1 {
+                        // turn animatin off
+                        self.wave_period = 0.0;
+                        return AnimationState::Finished;
+                    }
                 }
+                self.color * intensity
+            } else {
+                // solid
+                self.color
+            };
+            for led in frame {
+                *led = color;
             }
-            AnimationState::Running
-        } else {
-            AnimationState::Finished
         }
+        AnimationState::Running
```

Comment:
see if it makes sense to break a subset of this section out into a helper and write a unit test